ironfish/src/rpc/stream.ts

```diff
@@ -1,32 +1,66 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
-import { PromiseResolve } from '../utils'
+import { PromiseReject, PromiseResolve } from '../utils'
 
+/*
+ * A Stream takes a stream of data and transforms it into an iterable
+ * On one end the writer of the data is calling stream.write(data)
+ * and on the other end the consumer of the data is calling stream.next()
+ *
+ * If stream.next() is called before data is written then the request will be queued
+ * up in the `waiting` list by saving functions to resolve and reject the request
+ *
+ * stream.next()      -> [(res, rej)]               a resolve and reject for the data is queued in `waiting`
+ * stream.next()      -> [(res, rej), (res, rej)]
+ * stream.write(data) -> [(res, rej)]               the first Promise in waiting is resolved with the data
+ * stream.write(data) -> []                         the final Promise in waiting is resolved with the data
+ *
+ * If stream.write() is called before the reader requests with stream.next() then the
+ * data will be queued up in the `buffer` list
+ *
+ * stream.write(1) -> [1]                            data is queued up in the buffer
+ * stream.write(2) -> [1, 2]
+ * stream.write(3) -> [1, 2, 3]
+ * stream.next()   -> [2, 3]                         data is consumed from the buffer and resolved to next()
+ * stream.next()   -> [3]
+ *
+ * Because Stream implements AsyncIterable it can also be used in for...in loops
+ */
```

Comment:
I think large documentation like this is a bit strange. Do you need an explanation on how to use something as simple as this?

```
interface Stream<T> {
    write()
    next()
    close()
}
```

I read the docs but they kind of just re-describe how an async stream is implemented but that's already clear because the class is called Stream<T>. Is it because this is called Stream<T> and not AsyncStream<T>?

Reply:
Yea it was pretty confusing for me initially. Now that I understand it it makes sense but I figured this would help when we come back to it or someone new comes in. It's a simple interface but internal workings I don't think are clear just by looking at the interface. Like what happens when write and next are not called in step. I also think it's fine to provide a comment because this seems like a pretty reusable component that's unlikely to change going forward. If it was just a one-off pattern I wouldn't provide a comment.